### PR TITLE
Hide bottom bar

### PR DESCRIPTION
**Pull request by imdkhairul**
## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 10.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift [hunk 1]
```diff
@@ -31,6 +31,17 @@ public extension UINavigationController {
...
         CATransaction.commit()
     }
+    
+    /// SwifterSwift: Push ViewController with hide Bottom Bar.
+    ///
+    /// - Parameters:
+    ///   - viewController: viewController to push.
+    ///   - hideBottomBar: set this true parameter to hide bottom bar (like a tab bar).
+    
+    func pushViewController(_ viewController: UIViewController, hideBottomBar:Bool){
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift:41]:
```suggestion
    func pushViewController(_ viewController: UIViewController, hidesBottomBar: Bool, animated: Bool) {
```

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift [hunk 2]
```diff
@@ -53,5 +53,15 @@ final class UINavigationControllerExtensionsTests: XCTestCase {
         XCTAssertNotNil(color)
         XCTAssertEqual(color!, .red)
     }
+    
+    func testHideBottomBar(){
```
Comment (guykogus) [inline Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:57]:
```suggestion
    func testHideBottomBar() {
```

**Review by guykogus — CHANGES_REQUESTED** [review]

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift [hunk 3]
```diff
@@ -53,5 +53,15 @@ final class UINavigationControllerExtensionsTests: XCTestCase {
...
+        let rootVC = UIViewController()
+        let navigationController = UINavigationController(rootViewController: rootVC)
+        let vcToPush = UIViewController()
+        let tabVC = UITabBarController()
+        tabVC.viewControllers = [navigationController]
+        navigationController.pushViewController(vcToPush, hideBottomBar: true)
+        XCTAssertEqual(vcToPush.hidesBottomBarWhenPushed, true)
+    }
```
[the remark this answers is not in the document]
Comment (imdkhairul) [reply Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:65]: Sure

File: CHANGELOG.md [hunk 4]
```diff
@@ -18,6 +18,8 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - `init(randomOfLength:)` deprecated in favor of `String.random(ofLength:)`. [#1115](https://github.com/SwifterSwift/SwifterSwift/issues/1115) by [guykogus](https://github.com/guykogus)
 
 ### Added
+**UINavigationController**
```
Comment (guykogus) [inline CHANGELOG.md:21]:
```suggestion
- **UINavigationController**
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -18,6 +18,8 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
...
+  - Added `pushViewController(_:hidesBottomBar:animated:)` for hiding the bottom bar when pushing any `UIViewController`. [#1117](https://github.com/SwifterSwift/SwifterSwift/issues/1117) [imdkhairul](https://github.com/imdkhairul)
```
Comment (guykogus) [inline CHANGELOG.md:22]:
```suggestion
  - Added `pushViewController(_:hidesBottomBar:animated:)` for hiding the bottom bar when pushing any `UIViewController`. [#1117](https://github.com/SwifterSwift/SwifterSwift/issues/1117) by [imdkhairul](https://github.com/imdkhairul)
```

File: Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift [hunk 6]
```diff
@@ -31,6 +31,17 @@ public extension UINavigationController {
...
         CATransaction.commit()
     }
+    
+    /// SwifterSwift: Push ViewController with hide Bottom Bar.
+    ///
+    /// - Parameters:
+    ///   - viewController: viewController to push.
+    ///   - hideBottomBar: set this true parameter to hide bottom bar (like a tab bar).
+    
+    func pushViewController(_ viewController: UIViewController, hideBottomBar:Bool, animated:Bool) {
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift:41]: Please be aware of spaces and change from `hideBottomBar` to `hidesBottomBar`
```suggestion
    /// SwifterSwift: Pushes a view controller while hiding or showing the bottom bar.
    ///
    /// - Parameters:
    ///   - viewController: The view controller to push.
    ///   - hidesBottomBar: If `true`, hides the bottom bar (e.g. tab bar).
    ///   - animated: Specify `true` to animate the transition.
    func pushViewController(_ viewController: UIViewController, hidesBottomBar: Bool, animated: Bool) {
```

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift [hunk 7]
```diff
@@ -53,5 +53,19 @@ final class UINavigationControllerExtensionsTests: XCTestCase {
...
+        let navigationController = UINavigationController(rootViewController: rootVC)
+        let vcToPush = UIViewController()
+        let tabVC = UITabBarController()
+        tabVC.viewControllers = [navigationController]
+        navigationController.pushViewController(vcToPush, hideBottomBar: true, animated: true)
+        XCTAssert(vcToPush.hidesBottomBarWhenPushed)
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1){
+            XCTAssert(tabVC.tabBar.isHidden)
+        }
+
```
Comment (guykogus) [inline Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:68]: If you want to test something with a delay, you need to wait for an expectation.
Also, you probably don't even need the delay for the async call, if you set `animated: false`
```suggestion
        let expectation = XCTestExpectation(description: "Tab bar hidden")
        DispatchQueue.main.async {
            XCTAssert(tabVC.tabBar.isHidden)
            expectation.fulfill()
        }
        wait(for: [expectation], timeout: 0.1)
```

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift [hunk 8]
```diff
@@ -53,5 +53,19 @@ final class UINavigationControllerExtensionsTests: XCTestCase {
...
         XCTAssertEqual(color!, .red)
     }
+    
+    func testHideBottomBar() {
+        let rootVC = UIViewController()
+        let navigationController = UINavigationController(rootViewController: rootVC)
+        let vcToPush = UIViewController()
+        let tabVC = UITabBarController()
+        tabVC.viewControllers = [navigationController]
+        navigationController.pushViewController(vcToPush, hideBottomBar: true, animated: true)
```
Comment (guykogus) [inline Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:63]: Don't animate in unit tests.
```suggestion
        navigationController.pushViewController(vcToPush, hideBottomBar: true, animated: false)
```

**Review by guykogus — CHANGES_REQUESTED** [review]

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift (hunk 7 above)
Comment (guykogus) [reply Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:68]: Actually, if you don't animate then you may not need the `async` call altogether and can just run the test. Try that first.

File: Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift [hunk 9]
```diff
@@ -31,6 +31,18 @@ public extension UINavigationController {
...
     }
+    
+    /// SwifterSwift: Pushes a view controller while hiding or showing the bottom bar.
+    ///
+    /// - Parameters:
+    ///   - viewController: The view controller to push.
+    ///   - hidesBottomBar: If `true`, hides the bottom bar (e.g. tab bar).
+    ///   - animated: Specify `true` to animate the transition.
+    
+    func pushViewController(_ viewController: UIViewController, hidesBottomBar: Bool, animated: Bool) {
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift:42]: Remove extra empty line
```suggestion
    ///   - animated: Specify `true` to animate the transition.
    func pushViewController(_ viewController: UIViewController, hidesBottomBar: Bool, animated: Bool) {
```

**Review by guykogus — CHANGES_REQUESTED** [review]

File: Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift [hunk 10]
```diff
@@ -31,6 +31,17 @@ public extension UINavigationController {
         pushViewController(viewController, animated: true)
         CATransaction.commit()
     }
+    
+    /// SwifterSwift: Pushes a view controller while hiding or showing the bottom bar.
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift:35]:
```suggestion
#if !os(tvOS)
    /// SwifterSwift: Pushes a view controller while hiding or showing the bottom bar.
```

File: Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift [hunk 11]
```diff
@@ -31,6 +31,17 @@ public extension UINavigationController {
...
+    ///
+    /// - Parameters:
+    ///   - viewController: The view controller to push.
+    ///   - hidesBottomBar: If `true`, hides the bottom bar (e.g. tab bar).
+    ///   - animated: Specify `true` to animate the transition.
+    func pushViewController(_ viewController: UIViewController, hidesBottomBar: Bool, animated: Bool) {
+        viewController.hidesBottomBarWhenPushed = hidesBottomBar
+        pushViewController(viewController, animated: animated)
+    }
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UINavigationControllerExtensions.swift:44]:
```suggestion
    }
#endif
```

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift [hunk 12]
```diff
@@ -53,5 +53,16 @@ final class UINavigationControllerExtensionsTests: XCTestCase {
         XCTAssertNotNil(color)
         XCTAssertEqual(color!, .red)
     }
+    
+    func testHideBottomBar() {
```
Comment (guykogus) [inline Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:57]:
```suggestion
#if !os(tvOS)
    func testHideBottomBar() {
```

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift [hunk 13]
```diff
@@ -53,5 +53,16 @@ final class UINavigationControllerExtensionsTests: XCTestCase {
...
+        let rootVC = UIViewController()
+        let navigationController = UINavigationController(rootViewController: rootVC)
+        let vcToPush = UIViewController()
+        let tabVC = UITabBarController()
+        tabVC.viewControllers = [navigationController]
+        navigationController.pushViewController(vcToPush, hidesBottomBar: true, animated: false)
+        XCTAssert(vcToPush.hidesBottomBarWhenPushed)
+        XCTAssert(tabVC.tabBar.isHidden)
+    }
```
Comment (guykogus) [inline Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:66]:
```suggestion
    }
#endif
```

**Review by guykogus — CHANGES_REQUESTED** [review]

**Comment by guykogus** [comment]
@imdkhairul `hidesBottomBarWhenPushed` is not available for tvOS, so we need to disable the function and tests there.

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift (hunk 13 above)
Comment (imdkhairul) [reply Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:67]: @guykogus Disabled for tvOS

File: Tests/UIKitTests/UINavigationControllerExtensionsTests.swift (hunk 12 above)
Comment (imdkhairul) [reply Tests/UIKitTests/UINavigationControllerExtensionsTests.swift:58]: @guykogus Disabled for tvOS

**Review by guykogus — APPROVED** [review]

**Pull request merged**